src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,546 @@
+    try {
+      SubscriberStubSettings subscriberSettings = getSubscriberSettings();
+      subscriber = getSubscriber(subscriberSettings);
+    } catch (IOException ioe) {
+      //This exception is thrown when the subscriber could not be created.
+      //Report the exception and stop the receiver.
+      String message =
+        String.format(SUBSCRIBER_ERROR_MSG, subscription, project);
+      stop(message, ioe);
+      return;
+    }
+
+    executor.scheduleWithFixedDelay(this::receiveMessages, 0, 100, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Fetch new messages for our subscription.
+   * Implements exponential backoff strategy when a retryable exception is received.
+   * This method stops the receiver if a non retryable ApiException is thrown by the Google Cloud subscriber client.
+   */
+  protected void receiveMessages() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+
+    //Try with backoff until stopped or the task succeeds.
+    while (!isStopped()) {
+      try {
+        fetchAndAck();
+        return;
+      } catch (ApiException ae) {
+        // A retryable exception means the request can be safely retried at a later time.
+        // In this case, the exponential backoff logic starts increasing the delay for requests).
+
+        // A non-retryable exception means that we will not be able to pull any more messages from this topic.
+        // This can be related to credentials, the subscription getting deleted, or an error from the Pub/Sub service.
+        // In this case, the receiver gets restarted, credentials re-initialized, subscription recreated if needed,
+        // and then the receiver starts fetching messages once again.
+        if (isRetryable(ae)) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+        } else {
+          // Restart the receiver if the exception is not retryable.
+          String message =
+            String.format(FETCH_ERROR_MSG, subscription, project);
+          restart(message, ae);
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Fetch new messages, store in Spark's memory, and ack messages.
+   * Based on SubscribeSyncExample.java in Google's PubSub examples.
+   */
+  protected void fetchAndAck() {
+    //Get the maximun number of messages to get. If this number is less or equal than 0, do not fetch.
+    int maxMessages = bucket.get();
+    if (maxMessages <= 0) {
+      return;
+    }
+
+    PullRequest pullRequest =
+      PullRequest.newBuilder()
+        .setMaxMessages(maxMessages)
+        .setSubscription(subscription)
+        .build();
+    PullResponse pullResponse = subscriber.pullCallable().call(pullRequest);
+
+    List<ReceivedMessage> receivedMessages = pullResponse.getReceivedMessagesList();
+
+    //If there are no messages to process, continue.
+    if (receivedMessages.isEmpty()) {
+      return;
+    }
+
+    //Decrement number of available messages in bucket.
+    bucket.updateAndGet(x -> x - receivedMessages.size());
```

Comment:
Seems like you can simply use `decrementAndGet(receivedMessages.size())`

Reply:
I don't think AtomicInteger implements this method in JDK8, at least https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html